File: LogisticRegression.cpp
#include<Eigen/Dense>
#include<iostream>
#include<cassert>
#include<stdexcept>
class LogisticRegression{
private:
    Eigen::VectorXd W;
    double b;
    bool ridge = false;
    bool lasso = false;
    double lambdaL1 = 0.0;
    double lambdaL2 = 0.0;

public:
    LogisticRegression(int numFeatures, bool ridge_ = false, bool lasso_ = false,  double lambdaL1_ = 0.0, double lambdaL2_ = 0.0)
        : W(Eigen::VectorXd::Zero(numFeatures)),
        b(0.0f),
        ridge(ridge_),
        lasso(lasso_),
        lambdaL1(lambdaL1_),
        lambdaL2(lambdaL2_)
    {
        if((ridge || lasso) && (lambdaL1 <= 0.0 && lambdaL2 <= 0.0))
            throw std::invalid_argument("lambda must be > 0 when regualarization is enabled");
        if(!ridge && !lasso && (lambdaL1 != 0.0 || lambdaL2 != 0.0))
            throw std::invalid_argument("lambda given but no regularization enabled");
        if(ridge && lasso && (lambdaL1 <= 0.0 || lambdaL2 <= 0.0))
            throw std::invalid_argument("both lambdaL2 (ridge) and lambdaL1 (lasso) must be > 0 for elastic net");
        if(ridge && !lasso && lambdaL2 <= 0.0)
            throw std::invalid_argument("lambdaL2 must be > 0 for ridge");
        if(!ridge && lasso && lambdaL1 <= 0.0)
            throw std::invalid_argument("lambdaL1 must be > 0 for lasso");       
    }

    Eigen::VectorXd predict(const Eigen::MatrixXd& X) const{
        assert(X.cols() == W.size());
        Eigen::VectorXd z = X * W;
        z.array() += b;
        Eigen::VectorXd exponential = (-z.array()).exp();
        Eigen::VectorXd y_hat = (1.0/(1.0+exponential.array()));
        return y_hat;
    }

    double loss(const Eigen::VectorXd& y, const Eigen::VectorXd& y_hat) const{
        assert(y.size() == y_hat.size());
        double eps = 1e-15;

        double loss = -(
             y.array() * (y_hat.array() + eps).log()
             + (1.0 - y.array()) * (1.0 - y_hat.array() + eps).log() 
             ).mean();
        
        
        if (ridge)
            loss += W.squaredNorm() * (lambdaL2/(2.0 * y.size()));
        if (lasso)
            loss += W.lpNorm<1>() * (lambdaL1/y.size());

        return loss;
    }

    void train(const Eigen::MatrixXd& X, const Eigen::VectorXd& y, int numEpochs, double lr){        
        assert(X.cols() == W.size());
        assert(y.size() == X.rows());
        
        for(int i = 0; i < numEpochs; i++){
            Eigen::VectorXd y_hat = predict(X);

            Eigen::VectorXd dz = y_hat - y;
            double m = static_cast<double>(dz.size());
            Eigen::MatrixXd dW = (1/m) * (X.transpose() * dz);
            double db = (1/m) * dz.sum();
            if(ridge)
                dW += (lambdaL2/m) * W;
            if(lasso)
                dW += (lambdaL1/m) * W.array().sign().matrix();
            W.array() -= lr * dW.array();
            b -= lr * db;
        }
    }
};


int main(){
    Eigen::MatrixXd X(6,2);
    X << 1,2,
         2,3,
         3,1,
         6,7,
         7,8,
         8,6;
    Eigen::VectorXd Y(6);
    Y << 0,0,0,1,1,1;
    LogisticRegression Model(2,true, true, 0.01,0.01);
    Model.train(X,Y,100000,0.1);

    Eigen::VectorXd Y_hat = Model.predict(X);
    double loss = Model.loss(Y, Y_hat);

    std::cout << "Loss = " << loss << "\n";
    return 0;
}

File: LogisticRegression.py
import numpy as np

class LogisticRegression:
    def __init__(self,numFeatures:int, ridge_:bool = False, lasso_:bool = False, lambda_l1_:float  = 0.0, lambda_l2_ = 0.0):
        self.W = np.zeros(shape = (numFeatures), dtype = np.float64)
        self.b = 0.0
        self.ridge = ridge_
        self.lasso = lasso_
        self.lambda_l1 = lambda_l1_
        self.lambda_l2 = lambda_l2_
        if (self.ridge or self.lasso) and (self.lambda_l1 <= 0.0 and self.lambda_l2 <= 0.0):
            raise Exception("lambda must be > 0 when regularization is enabled")
        if not self.ridge and not self.lasso and (self.lambda_l1 != 0.0 and self.lambda_l2 != 0.0):
            raise Exception("lambda given but no regularization enabled");
        if(self.ridge and self.lasso and (self.lambda_l1 <= 0.0 or self.lambda_l2 <= 0.0)):
            raise Exception("both lambda_l2 (ridge) and lambda_l1 (lasso) must be > 0 for elastic net")
        if(self.ridge and not self.lasso and self.lambda_l2 <= 0.0):
            raise Exception("lambda_l2 must be > 0 for ridge")
        if(not self.ridge and self.lasso and self.lambda_l1 <= 0.0):
            raise Exception("lambda_l1 must be > 0 for lasso")

        
    def predict(self,X:np.ndarray):
        assert X.shape[1] == self.W.size

        z = X @ self.W + self.b
        y_hat = 1 / (1 + np.exp(-z))
        return y_hat

    def loss(self, y:np.ndarray, y_hat:np.ndarray):
        assert y.shape == y_hat.shape
        
        loss = -np.mean(y * np.log(y_hat + 1e-15) + (1-y) * np.log(1 - y_hat + 1e-15) ) 

        
        if self.ridge:
            loss += (self.lambda_l2/(2 * y.size)) * np.sum(self.W ** 2)

        if self.lasso:
            loss += (self.lambda_l1/y.size) * np.sum(np.abs(self.W))


        return loss

    def train(self, X:np.ndarray, y:np.ndarray, num_epochs:int, lr:float):
        assert X.shape[0] == y.size 
        assert X.shape[1] == self.W.size

        for i in range(num_epochs):
            y_hat = self.predict(X)
            
            dz = y_hat - y
            dW = (1/dz.size) * (X.T  @ dz)
            db = (1/dz.size) * np.sum(dz)
            
            if self.ridge:
                dW += (self.lambda_l2/dz.size) * self.W 
            if self.lasso:
                dW += (self.lambda_l1/dz.size) * np.sign(self.W)

            self.W -= lr * dW
            self.b -= lr * db

X = np.array([
    [1, 2],
    [2, 3],
    [3, 1],
    [6, 7],
    [7, 8],
    [8, 6]
])
y = np.array([0,0,0,1,1,1])

model = LogisticRegression(2, True, True, 0.01, 0.01)
model.train(X, y, 100000, 2)
y_hat = model.predict(X)
loss = model.loss(y, y_hat)
print("loss = ", loss)
